### PR TITLE
Catalog: expose `location` for namespaces, iam-policy per location

### LINKED ISSUE #8740: [Catalog] Security: Harden s3-session-credentials
Similar to #8739, the client should only be able to operate within the table's locations.

I suspect, this is possible via S3 session-credentials policy.
Would need to issue new session-credentials for every table access though.

Not sure how good that might work though.

### PR DESCRIPTION
**Pull request by snazy**
This change heavily changes how locations are handled and how S3 request signing and IAM policies work.

The result is:

* Iceberg's `write.object-storage.enabled=true` is respected - in Nessie it's sufficient to just set that single property.
* S3 request signing:
  * User's provilege (write or read) for a table is respected
  * Object-storage friendly storage-uris are supported
  * Writing is only possbile if it happens against the "current" warehouse
  * S3 signing endpoints verify the integrity of the S3 signing parameters using a cryptographic algorithm
* IAM policies:
  * User's provilege (write or read) for a table is respected
  * Object-storage friendly storage-uris are supported
  * Policies are dynamically generated, but can be enhanced with additional expressions or completely replaced with a custom static policy
* Request signing and assume-role/vended-credentials properties are now returned from a "load table" Iceberg REST call, but no longer from the catalog's `/v1/config` endpoint. The functionality depends on the S3 bucket configuration options `request-signing-enabled` and `assume-role-enabled`.
* Works with Trino (more Trino stuff in a follow-up)
  * Exposes the `location` property for namespaces
* The `X-Iceberg-Access-Delegation` request header is respect, according to Nessie's configuration

A bunch of configurations related to S3 buckets had to be changed. S3 buckets now support both request-signing and assume-role/credentials-vending. To allow this, the following configuration options have _changed_:

* The S3 bucket configuration option `client-authentication-mode` has been removed (defaulted to `REQUEST_SIGNING`).
* A new S3 bucket configuration option `request-signing-enabled` has been added (defaults to `true`).

The assume-role/credentials-vending configuration has been split into 2 sets: one for Nessie server credentials and one for client credentials. This allows the Nessie server to directly use the configured S3 credentials, but hand out scoped-down credentials from STS to clients.

* Nessie server IAM options are configured via the `.server-iam.` prefix.
* Nessie server IAM must be enabled using `.server-iam.enabled=true` on either the default S3 options or bucket-specific options.
* Client IAM must be enabled using `.client-iam.enabled=true` on either the default S3 options or bucket-specific options.
* Inheritance of default / bucket-specific options is supported for both `.server-iam.*` and `.client-iam.*` configuration options.

The client IAM policy can be provided using either a static policy via `.client-iam.policy`, or, if `.client-iam.policy` is not present, Nessie will generate the policy scoped down to a table's location.  Additional IAM policy statements can be supplied via the `.client-iam.client-iam-statements` _list_ property.

Fixes #8740

### CODE REVIEW

**Comment by olivier-derom** [comment]
Does this change mean that in Trino we can simply run 'CREATE SCHEMA myschema' without the 'WITH (location=s3://...)' property?
I assume that this property is then automatically generated by Nessie based on the warehouse configuration in nessie, essentially putting all the location responsibilities on Nessie and not the user/client/trino.

In such case, what is the behavior in case we do manually provide the location in a namespace/schema? or if its not done in the schema, but at table level?

And would the (optional) 'iceberg.rest-catalog.warehouse' property in the trino connection have any effect?
[https://trino.io/docs/current/object-storage/metastores.html#iceberg-rest-catalog](https://trino.io/docs/current/object-storage/metastores.html#iceberg-rest-catalog)

**Comment by snazy** [comment]
The _current state_ of this (big) PR makes life w/ Trino easier and also adds support for `write.object-storage.enabled=true`.

> Does this change mean that in Trino we can simply run 'CREATE SCHEMA myschema' without the 'WITH (location=s3://...)' property? I assume that this property is then automatically generated by Nessie based on the warehouse configuration in nessie, essentially putting all the location responsibilities on Nessie and not the user/client/trino.

The `location` is actually not required by Nessie. [Here](https://github.com/snazy/nessie/blob/namespace-location/servers/quarkus-server/src/intTest/java/org/projectnessie/server/catalog/AbstractTrino.java) are the tests that are working fine against Trino with this PR. The SQL statements should help answer the questions.

> In such case, what is the behavior in case we do manually provide the location in a namespace/schema? or if its not done in the schema, but at table level?

Nessie still controls the `location` for (new) tables to somewhat ensure that the locations don't overlap (and you gain access to otherwise protected data). The whole story isn't yet perfect, but an effort towards that goal.

> And would the (optional) 'iceberg.rest-catalog.warehouse' property in the trino connection have any effect?

The warehouse is passed down to Nessie, so Nessie knows the warehouse you wanna operate against, if it's not the default warehouse.

File: catalog/service/rest/build.gradle.kts [hunk 1]
```diff
@@ -45,6 +45,7 @@ dependencies {
   implementation(enforcedPlatform(libs.quarkus.bom))
   implementation("io.quarkus.resteasy.reactive:resteasy-reactive-common")
   implementation("io.quarkus.resteasy.reactive:resteasy-reactive")
+  implementation("org.eclipse.microprofile.config:microprofile-config-api")
```
[the remark this answers is not in the document]
Comment (snazy) [reply catalog/service/rest/build.gradle.kts:48]: Hm - leftover.

File: catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/IcebergApiV1TableResource.java [hunk 2]
```diff
@@ -167,23 +180,66 @@ R loadTableResultFromSnapshotResponse(
...
+            nessieSnapshot,
+            warehouseLocation,
+            tableMetadata.location(),
+            properties,
+            prefix,
+            contentKey,
+            dataAccess,
+            writeAccessGranted);
+
+    IcebergTableMetadata metadataTweak =
```
[the remark this answers is not in the document]
Comment (snazy) [reply catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/IcebergApiV1TableResource.java:236]: Refactoring this a bit to make it clearer.

File: helm/nessie/templates/_helpers.tpl [hunk 3]
```diff
@@ -158,15 +158,29 @@ Apply S3 catalog options.
...
+{{- if .clientIam.statements -}}
+{{- range $i, $statement := .clientIam.statements -}}
+{{- $_ := set $map ( print $prefix "client-iam.statements[%d]" $i ) $statement -}}
+{{- end -}}
+{{- end -}}
+{{- end -}}
+{{- if .serverIam -}}
+{{- if .serverIam.enabled -}}{{- $_ := set $map ( print $prefix "server-iam.enabled" ) .serverIam.enabled -}}{{- end -}}
+{{- if .serverIam.policy -}}{{- $_ := set $map ( print $prefix "server-iam.policy" ) .serverIam.policy -}}{{- end -}}
+{{- if .serverIam.roleArn -}}{{- $_ := set $map ( print $prefix "aserver-iam.ssume-role" ) .serverIam.roleArn -}}{{- end -}}
```
Comment (adutra) [inline helm/nessie/templates/_helpers.tpl:180]:
```suggestion
{{- if .serverIam.roleArn -}}{{- $_ := set $map ( print $prefix "server-iam.ssume-role" ) .serverIam.roleArn -}}{{- end -}}
```

File: servers/quarkus-config/src/main/java/org/projectnessie/quarkus/config/CatalogS3BucketConfig.java [hunk 4]
```diff
@@ -17,17 +17,12 @@
...
 import io.smallrye.config.WithName;
 import java.util.Optional;
+import org.projectnessie.catalog.files.s3.S3AuthType;
 import org.projectnessie.catalog.files.s3.S3BucketOptions;
-import org.projectnessie.catalog.files.s3.S3ClientAuthenticationMode;
-import org.projectnessie.catalog.files.s3.S3ServerAuthenticationMode;
 
 public interface CatalogS3BucketConfig extends S3BucketOptions {
 
   @WithName("server-auth-mode")
```
Comment (adutra) [inline servers/quarkus-config/src/main/java/org/projectnessie/quarkus/config/CatalogS3BucketConfig.java:25]: We could imho even remove the method override completely wdyt?

```suggestion
  @WithName("auth-type")
```

**Review by adutra — APPROVED** [review]

**Pull request merged**